lib/graphlib/graphlib_graph.ml

```diff
@@ -186,10 +186,123 @@ module Partition = struct
     let find x = Option.Monad_infix.(Hashtbl.find comps x >>= find_root) in
     {roots; groups; find}
 
+  let equiv t x y = Option.equal Equiv.equal (t.find x) (t.find y)
+
+  (* The trivial partition with a single class, or zero if elts is empty *)
+  let trivial elts =
+    let head = Set.choose elts in
+    match head with
+    | None ->
+       let roots = [||] in
+       let groups = [||] in
+       let find _ = None in
+       {roots; groups; find}
+    | Some h ->
+       let roots = Array.create ~len:1 h in
+       let groups = Array.create ~len:1 (create_set elts) in
+       let find x = if Set.mem elts x then Some 0 else None in
+       {roots; groups; find}
+
+  (* The discrete partition with one class per element *)
+  let discrete elts =
+    let comparator = Set.comparator elts in
+    let {Comparator.compare} = comparator in
+    (* Produces a sorted array per the spec *)
+    let roots = Set.to_array elts in
+    let groups = elts |> Set.to_array |>
+                   Array.map ~f:(fun x -> x |> Set.singleton ~comparator |> create_set)
+    in
+    let find x = Array.binary_search roots ~compare `First_equal_to x in
+    {roots; groups; find}
+
+  (* Takes a partition and a congruence and splits each equivalence class into
+     elements related by the congruence.
+     Takes in a comparison function to test for membership in each class.
+   *)
+  let refine t ~rel ~comp =
+    let refine_group g =
+      let s = g#enum in
+      let rec insert x seqs =
+        match seqs with
+        | [] -> [Seq.return x]
+        | s::ss ->
+           let y = Seq.hd_exn s in
+           if rel x y = 0 then
+             Seq.cons x s :: ss
+           else s :: (insert x ss)
+      in
+      let new_enums = Seq.fold s ~init:[] ~f:(fun seqs x ->
+                         insert x seqs)
+      in
+      let mk_set s =
+      object
+        method enum = s
+        method mem x = Seq.mem s x (fun x y -> comp x y = 0)
+      end
+      in
+      List.map new_enums ~f:mk_set
+    in
+    let groups_list = Array.fold t.groups ~init:[] ~f:(fun seqs g -> refine_group g @ seqs) in
+    let groups = Array.of_list groups_list in
+    Array.sort groups ~cmp:(fun s1 s2 ->
+        let h1 = Seq.hd_exn s1#enum in
+        let h2 = Seq.hd_exn s2#enum in
+        comp h1 h2);
+    let roots = Array.map ~f:(fun s -> Seq.hd_exn s#enum) groups in
+    let find x = Array.binary_search roots ~compare:comp `First_equal_to x in
+    {roots; groups; find}
+    
+  (* Take two elements and combine their classes if both have a class, 
+     do nothing otherwise *)
+  let merge t x y =
+    (* Assuming i < j,
+       create a new array a', identical to a execept that
+       a'[i] = x
+       a'[j] = a[j+1]
+       a'[j+1] = a[j+2]
+       ...
+     *)
+    let array_replace a i j x =
+      assert (i < j && Array.length a > 0);
```

Comment:
This should violate the fact that `~equiv t x y` holds, since we only call `array_replace` in the else branch of an `equiv t x y` test.

Maybe this deserves a comment, or `array_replace` can be made more robust...

Reply:
Oh, I missed that equiv test, indeed, you're right.